### PR TITLE
[apache] [jax-rs2] Instrumentation specific multiple span reporting Issue

### PR DESCRIPTION
**Pull request by dray92**
- Apache span interceptors have lighter dependency on threadlocal; newly created span is not pushed
- HttpContext is used to propagate the span between interceptors
- Similar changes in jax-rs2 instrumentation to reduce dependence on thread local, and instead use jax-rs2 specific `javax.ws.rs.client.ClientRequestContext`

Signed-off-by: Debosmit Ray <debo@uber.com>

### CODE REVIEW

File: jaeger-core/src/main/java/com/uber/jaeger/Span.java [hunk 1]
```diff
@@ -147,6 +148,10 @@ public SpanContext context() {
 
   @Override
   public void finish() {
+    if (finished) {
```
Comment (yurishkuro) [inline jaeger-core/src/main/java/com/uber/jaeger/Span.java]: this logic should be in the `finishWithDuration` method, and inside the synchronized block

**Comment by vprithvi** [comment]
Could you split this into two separate PRs by separating the instrumentation fixes from the changes in the Span class?

**Comment by dray92** [comment]
@vprithvi definitely -- lemme just remove the span class changes

File: jaeger-apachehttpclient/src/test/java/com/uber/jaeger/httpclient/JaegerRequestAndResponseInterceptorIntegrationTest.java [hunk 2]
```diff
@@ -124,7 +130,7 @@ public void testHttpClientTracing() throws Exception {
   private void verifyTracing(Span parentSpan) {
     //Assert that traces are correctly emitted by the client
     List<Span> spans = reporter.getSpans();
-    assertEquals(3, spans.size());
+    assertEquals(2, spans.size());
```
Comment (yurishkuro) [inline jaeger-apachehttpclient/src/test/java/com/uber/jaeger/httpclient/JaegerRequestAndResponseInterceptorIntegrationTest.java]: what is the reason for this change?

Comment (dray92) [reply jaeger-apachehttpclient/src/test/java/com/uber/jaeger/httpclient/JaegerRequestAndResponseInterceptorIntegrationTest.java]: Previously, when `pop()` was getting called, it would make another call to report the span. Hence the reporter was getting an extra span, and the last 2 spans were identical. This changes proves that the core issue is fixed.
